src/main/scala/com/yotpo/metorikku/test/Tester.scala

```diff
@@ -85,75 +91,273 @@ case class Tester(config: TesterConfig) {
           .outputMode(outputMode)
           .start()
           .processAllAvailable()
-        sparkSession.table(outputTableName).collect()
+        sparkSession.table(outputTableName)
       }
-      case false => df.collect()
+      case false => df
     }
   }
 
+
   private def compareActualToExpected(metricName: String): Array[String] = {
     var errors = Array[String]()
-    var errorsIndexArr = Seq[Int]()
     val metricExpectedTests = config.test.tests
+    val configuredKeys = config.test.keys
+    val allColsKeys = metricExpectedTests.mapValues(v=>v(0).keys.toList)
+    val keys = fillDefaultConfiguredKeys(configuredKeys, allColsKeys)
 
     metricExpectedTests.keys.foreach(tableName => {
-      val metricActualResultRows = extractTableContents(job.sparkSession, tableName, config.test.outputMode.get)
-      var metricExpectedResultRows = metricExpectedTests(tableName)
-      if (metricExpectedResultRows.length == metricActualResultRows.length) {
-        for ((metricActualResultRow, rowIndex) <- metricActualResultRows.zipWithIndex) {
-          val mapOfActualRow = metricActualResultRow.getValuesMap(metricActualResultRow.schema.fieldNames)
-          val matchingExpectedMetric = matchExpectedRow(mapOfActualRow, metricExpectedResultRows)
-          if (Option(matchingExpectedMetric).isEmpty) {
-            errorsIndexArr = errorsIndexArr :+ rowIndex
-            errors = errors :+ s"[$metricName - $tableName] failed on row ${rowIndex + 1}: " +
-              s"Didn't find any row in test_settings.json that matches ${mapOfActualRow}"
-          }
-          else {
-            metricExpectedResultRows = metricExpectedResultRows.filter(_ != matchingExpectedMetric)
+      var tableErrors = Array[String]()
+      var errorsIndexArr = Seq[Int]()
+      currentTableName = tableName
+      val actualResults = extractTableContents(job.sparkSession, tableName, config.test.outputMode.get)
+      val expectedResults = metricExpectedTests(tableName)
+      val tableKeys = keys(tableName)
+
+      val actualKeysList: _root_.scala.Array[_root_.java.lang.String] = getKeyListFromDF(actualResults, tableKeys).sorted
+      val expKeysList: _root_.scala.Array[_root_.java.lang.String] = getKeyListFromMap(expectedResults, tableKeys).sorted
+      if (expKeysList.deep != actualKeysList.deep) {
+        // compare keys with duplications count
+        tableErrors = tableErrors ++ compareKeys(expKeysList, actualKeysList, metricName, tableName, tableKeys)
+      }
+      val sortedExpectedResults = expectedResults.sortWith(sortRows)
```

Comment:
Print out the subset with errors, and quit. No point in continuing the test in case you found issues.